### PR TITLE
Migrate from Manifest v2 to Manifest v3

### LINKED ISSUE #11: Investigate (& maybe implement) upgrade to manifest v3
Mozilla & Google have released v3 of their addon/extension manifest. It would be good to see what exactly needs doing to this repo to make it compatible on v3 for both Chrome & Firefox.

https://blog.mozilla.org/addons/2021/05/27/manifest-v3-update

Goal is to maintain cross-browser support for this extension. Currently, Checka11y.css browser extension is compatible on both Firefox & Chrome using the same code (not all extensions can do this and they have to create 2 separate repos for each browser) - we want to try and maintain this if we can so there is less code to maintain.

### PR DESCRIPTION
**Pull request by tannerdolby**
Fixes #11 

Proposed Changes:

- Change `manifest_version` from 2 to 3 in the manifest
- Replace `browser_action` and/or `page_action` attributes with `action` as they are unified into a single `action` property in MV3
- Replace `chrome.browserAction` and/or `chrome.pageAction` with `chrome.action` as these are also unified into a single API in MV3
- Todo: Double check the MV3 migration checklist and the [Firefox MV3 update](https://blog.mozilla.org/addons/2021/05/27/manifest-v3-update/) to ensure nothing is missed. Otherwise, this might be able to stay as a "single" extension which serves both browsers like you've setup.

References:
MV3 Migration Checklist - https://developer.chrome.com/docs/extensions/mv3/mv3-migration-checklist/
Firefox MV3 Update - https://blog.mozilla.org/addons/2021/05/27/manifest-v3-update/

### CODE REVIEW

**Comment by jackdomleo7** [comment]
Thanks for this PR @tannerdolby! Later on, I'll test on Firefox & Chrome to ensure it still works. But based on your changes, I can't see any issues arising. I also can't see any issues arising when I submit new versions to Mozilla and Google (because they tend to be strict when reviewing browser extensions).

**Comment by jackdomleo7** [comment]
I tested this on Chrome and it worked perfectly (albeit some errors in the Chrome extension section that I believe are pre-existing). However, I tested this on Firefox and got this error:
image
Which of all errors, I was not expecting this error, especially since Mozilla sent me an email, as an add-on maintainer, inviting me to use the v3 of the manifest.

After reading the [blog post](https://blog.mozilla.org/addons/2021/05/27/manifest-v3-update/) again & more thoroughly, I just noticed that Mozilla will not be accepting manifest v3 submissions until early 2022.

With this in mind, I will keep an eye on the [MDN docs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/manifest_version) for this next year and I will revisit this PR at a later date.

Thanks again for this investigation and PR.

**Comment by jackdomleo7** [comment]
@tannerdolby Again, thanks for your work on this. I tested again today and still Firefox is showing me the same error about not supporting manifest v3. I will revisit this again towards the end of 2022.

More info: https://blog.mozilla.org/addons/2022/05/18/manifest-v3-in-firefox-recap-next-steps/

**Comment by tannerdolby** [comment]
@jackdomleo7 No problem! :) I'm always happy to help. This extension is really useful so I will always be around if you need any assistance.

I had a look at https://blog.mozilla.org/addons/2022/05/18/manifest-v3-in-firefox-recap-next-steps/ and like you stated it seems all we can do is revisit this towards the end of 2022 as Mozilla's team say they expect to launch MV3 support for all users by the end of this year.

**Comment by jackdomleo7** [comment]
Hi @tannerdolby it seems we might be able to move this forward seeing since Chrome supports manifest v3 and it seems Firefox now supports manifest v3 too

https://blog.mozilla.org/addons/2022/11/17/manifest-v3-signing-available-november-21-on-firefox-nightly/

**Comment by tannerdolby** [comment]
👋 @jackdomleo7 

Yeah, it appears this will be safe to move forward with now. After reading the article you linked, I see Firefox 109 should support MV3 and they will begin deprecating MV2 towards the end of 2023 depending on the MV3 rollout this year. I'm not sure anything needs changing for this PR, but let me know!

**Comment by jackdomleo7** [comment]
I don't think you need to do anything either, but I can't merge your PR because it says there are merge conflicts and/or the branch no longer exists, and I want to ensure you get the credit for this work 🙂

In your opinion, would it be a good idea to merge this ASAP?

**Comment by tannerdolby** [comment]
Thank you 🙂 Ah, yes I must've deleted my fork while doing "house cleaning" of my repos earlier this year. I've re-forked and will restore the branch.

I don't think we need to rush to merge this, especially since Firefox still supports MV2. I believe the original changes in this PR were the only breaking changes that needed to be addressed, but let's test all that we can to make sure its fully functional with the MV3 updates before merging.

**Comment by jackdomleo7** [comment]
Hi @tannerdolby, I think we can merge this whenever. I keep getting email threats from Google to update to v3. With Firefox, doesn't seem like they're in a rush but I don't have any active Firefox users anyway.

When you can, sometime before June 2024, could you resolve any conflicts and we'll get this merged and published.

Many thanks!

**Review by jackdomleo7 — APPROVED** [review]

**Review by jackdomleo7 — DISMISSED** [review]

**Pull request closed without merging**
